scala/private/rule_impls.bzl

```diff
@@ -548,22 +548,32 @@ def _create_scala_compilation_provider(ctx, ijar, source_jar, deps_providers):
         runtime_deps = runtime_deps,
     )
 
-def build_deployable(ctx, jars_list):
-    # This calls bazels singlejar utility.
-    # For a full list of available command line options see:
-    # https://github.com/bazelbuild/bazel/blob/master/src/java_tools/singlejar/java/com/google/devtools/build/singlejar/SingleJar.java#L311
-    # Use --compression to reduce size of deploy jars.
+def merge_jars(actions, deploy_jar, singlejar_executable, label, jars_list, main_class = ""):
+    """Calls Bazel's singlejar utility.
+
+    For a full list of available command line options see:
+    https://github.com/bazelbuild/bazel/blob/697d219526bffbecd29f29b402c9122ec5d9f2ee/src/java_tools/singlejar/java/com/google/devtools/build/singlejar/SingleJar.java#L337
+    Use --compression to reduce size of deploy jars.
+
+    Args:
+        actions: The actions module from ctx: https://docs.bazel.build/versions/master/skylark/lib/actions.html
+        deploy_jar: The deploy jar, usually defined in ctx.outputs.
+        singlejar_executable: The singlejar executable file.
+        label: The label of the target, usually from ctx.label.
+        jars_list: The jars to pass to singlejar.
+        main_class: The main class to run, if any. Defaults to an empty string.
+    """
     args = ["--compression", "--normalize", "--sources"]
     args.extend([j.path for j in jars_list])
-    if getattr(ctx.attr, "main_class", ""):
-        args.extend(["--main_class", ctx.attr.main_class])
-    args.extend(["--output", ctx.outputs.deploy_jar.path])
-    ctx.actions.run(
+    if main_class:
+        args.extend(["--main_class", main_class])
+    args.extend(["--output", deploy_jar.path])
+    actions.run(
         inputs = jars_list,
-        outputs = [ctx.outputs.deploy_jar],
-        executable = ctx.executable._singlejar,
+        outputs = [deploy_jar],
+        executable = singlejar_executable,
         mnemonic = "ScalaDeployJar",
-        progress_message = "scala deployable %s" % ctx.label,
+        progress_message = "scala deployable %s" % label,
```

Comment:
maybe rather than taking the label, take the full message. So we could do `"scala_binary jar: %s" % ctx.label` or `"scala_library jar: " + ctx.label` or something.

Because, really this function has nothing to do with scala anymore. It is really just a jar merging rule.